internal/zinc-core/src/main/scala/sbt/internal/inc/IncrementalCommon.scala

```diff
@@ -613,33 +658,33 @@ object IncrementalCommon {
    * @param equivS An equivalence function to compare stamps.
    * @return
    */
-  def isBinaryModified(
+  def isLibraryModified(
       skipClasspathLookup: Boolean,
       lookup: Lookup,
       previousStamps: Stamps,
       currentStamps: ReadStamps,
       previousRelations: Relations,
+      converter: FileConverter,
       log: Logger
-  )(implicit equivS: Equiv[XStamp]): File => Boolean = { (binaryFile: File) =>
+  )(implicit equivS: Equiv[XStamp]): VirtualFileRef => Boolean = { (binaryFile: VirtualFileRef) =>
     {
       def invalidateBinary(reason: String): Boolean = {
         log.debug(s"Invalidating '$binaryFile' because $reason"); true
       }
 
-      def compareStamps(previousFile: File, currentFile: File): Boolean = {
-        val previousStamp = previousStamps.binary(previousFile)
-        val currentStamp = currentStamps.binary(currentFile)
+      def compareStamps(previousFile: VirtualFileRef, currentFile: VirtualFileRef): Boolean = {
+        val previousStamp = previousStamps.library(previousFile)
+        val currentStamp = currentStamps.library(currentFile)
         if (equivS.equiv(previousStamp, currentStamp)) false
         else invalidateBinary(s"$previousFile ($previousStamp) != $currentFile ($currentStamp)")
       }
 
-      def isBinaryChanged(file: File): Boolean = {
-        def compareOriginClassFile(className: String, classpathEntry: File): Boolean = {
-          val resolved = Locate.resolve(classpathEntry, className)
-          val resolvedCanonical = resolved.getCanonicalPath
-          if (resolvedCanonical != binaryFile.getCanonicalPath)
-            invalidateBinary(s"${className} is now provided by ${resolvedCanonical}")
-          else compareStamps(binaryFile, resolved)
+      def isLibraryChanged(file: VirtualFileRef): Boolean = {
+        def compareOriginClassFile(className: String, classpathEntry: VirtualFileRef): Boolean = {
+          if (classpathEntry.id.endsWith(".jar") &&
+              (converter.toPath(classpathEntry).toString != converter.toPath(file).toString))
+            invalidateBinary(s"${className} is now provided by ${classpathEntry}")
+          else compareStamps(file, classpathEntry)
```

Comment:
This line seems to contain a bug: it compares the stamp of a class file to a stamp of a classpath entry (directory). Unsurprisingly this always fails:
```
Invalidating '.../build/classes/scala/main/.../Foo.class' because .../build/classes/scala/main/....class (farm(77c283a23fed88f2)) != .../build/classes/scala/main (absent)  <= RHS is the classes ***directory***.
```
The `resolve` step was completely dropped in this PR.

Reply:
I've just found out this bug and I am going to fix it.